src/core/src/UrlAssetAccessor.cpp

```diff
@@ -0,0 +1,268 @@
+    return cnt;
+}
+} //namespace cesium::omniverse
+
+extern "C" size_t headerCallback(char* buffer, size_t size, size_t nitems, void* userData) {
+    return cesium::omniverse::UrlAssetResponse::headerCallback(buffer, size, nitems, userData);
+}
+
+extern "C" size_t dataCallback(char* buffer, size_t size, size_t nitems, void* userData) {
+    return cesium::omniverse::UrlAssetResponse::dataCallback(buffer, size, nitems, userData);
+}
+
+namespace cesium::omniverse {
+void UrlAssetResponse::setCallbacks(CURL* curl) {
+    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, ::dataCallback);
+    curl_easy_setopt(curl, CURLOPT_WRITEDATA, this);
+    curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, ::headerCallback);
+    curl_easy_setopt(curl, CURLOPT_HEADERDATA, this);
+}
+
+UrlAssetAccessor::UrlAssetAccessor(const std::filesystem::path& certificatePath)
+    : userAgent("Mozilla/5.0 Cesium for Omniverse")
+    , _certificatePath(certificatePath.generic_string()) {
+    // XXX Do we need to worry about the thread safety problems with this?
+    curl_global_init(CURL_GLOBAL_ALL);
+}
+
+UrlAssetAccessor::~UrlAssetAccessor() {
+    curl_global_cleanup();
+}
+
+curl_slist*
+UrlAssetAccessor::setCommonOptions(CURL* curl, const std::string& url, const CesiumAsync::HttpHeaders& headers) {
+    curl_easy_setopt(curl, CURLOPT_USERAGENT, userAgent.c_str());
+    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
+    if (!_certificatePath.empty()) {
+        curl_easy_setopt(curl, CURLOPT_CAINFO, _certificatePath.c_str());
+    }
+    curl_easy_setopt(curl, CURLOPT_SSL_OPTIONS, CURLSSLOPT_NATIVE_CA);
+    curl_easy_setopt(curl, CURLOPT_ACCEPT_ENCODING, "");
+    curl_easy_setopt(curl, CURLOPT_BUFFERSIZE, CURL_BUFFERSIZE);
+    curl_easy_setopt(curl, CURLOPT_MAXCONNECTS, 20L);
+    curl_easy_setopt(curl, CURLOPT_DNS_CACHE_TIMEOUT, 300L);
+    // curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 0L);
+    curl_slist* list = nullptr;
+    for (const auto& header : headers) {
+        std::string fullHeader = header.first + ":" + header.second;
+        list = curl_slist_append(list, fullHeader.c_str());
+    }
+    if (list) {
+        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, list);
+    }
+    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
+    return list;
+}
+
+CesiumAsync::Future<std::shared_ptr<CesiumAsync::IAssetRequest>> UrlAssetAccessor::get(
+    const CesiumAsync::AsyncSystem& asyncSystem,
+    const std::string& url,
+    const std::vector<CesiumAsync::IAssetAccessor::THeader>& headers) {
+    return asyncSystem.createFuture<std::shared_ptr<CesiumAsync::IAssetRequest>>([&](const auto& promise) {
+        std::shared_ptr<UrlAssetRequest> request = std::make_shared<UrlAssetRequest>("GET", url, headers);
+        asyncSystem.runInWorkerThread([promise, request, this]() {
+            CESIUM_TRACE("UrlAssetAccessor::get");
+            CurlHandle curl(this);
+            curl_slist* list = setCommonOptions(curl(), request->url(), request->headers());
```

Comment:
What's the overhead of calling `setCommonOptions`? Many of the options won't change per-request. Could some of them be set when acquiring a new handle?

If there's minimal overhead then no need to change it.

Reply:
I believe that it is harmless to set the options each time and that it's not too obvious from the docs what options persist after the curl_easy_reset() call. Merits further examination.